### PR TITLE
Add deployless multicall3

### PR DESCRIPTION
**Pull request by not-reed**
From the discussion https://github.com/wagmi-dev/viem/discussions/183

This introduces a new 'deployless' feature of multicall which allows for quering historic blocks prior to multicall3 deployment

---

### Detailed summary
- Added support for deployless multicall3
- Added `unstable_deployless` option to `createClient` and `multicall` functions
- Updated `deploylessMulticall3Abi` and `deploylessMulticall3Bytecode` constants

> The following files were skipped due to too many changes: `src/actions/public/multicall.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Review by witem — APPROVED** [review]

File: site/docs/contract/multicall.md [hunk 1]
```diff
@@ -291,6 +291,27 @@ const results = await publicClient.multicall({
 })
 ```
 
+### deployless (optional)
```
Comment (jxom) [inline site/docs/contract/multicall.md:294]: Wonder if this option could be enabled [globally](https://viem.sh/docs/clients/public.html#batch-multicall-batchsize-optional) as well (ie. `batch.multicall.deployless`)?

File: src/actions/public/multicall.ts [hunk 2]
```diff
@@ -276,3 +262,81 @@ export async function multicall<
...
+        : { data, blockNumber, blockTag }
+
+      const callReturn = await getAction(client, call)(params)
+
+      const outputItem = getAbiItem({ abi: multicall3Abi, name: 'aggregate3' })
+
+      if (callReturn.data) {
+        const values = decodeAbiParameters(outputItem.outputs, callReturn.data)
+        return values[0] as any
+      }
```
Comment (jxom) [inline src/actions/public/multicall.ts:339]: Can we use `decodeFunctionResult` here?

**Review by jxom — COMMENTED** [review]
Looks nice so far! A few comments.

File: site/docs/contract/multicall.md (hunk 1 above)
Comment (not-reed) [reply site/docs/contract/multicall.md:294]: good call :+1: done

File: src/actions/public/multicall.ts (hunk 2 above)
Comment (not-reed) [reply src/actions/public/multicall.ts]: nice, yeah :+1:

**Comment by not-reed** [comment]
@jxom I was just wondering if there was any update to this :)

**Comment by witem** [comment]
Hi @jxom 
any updates related to this PR?

**Comment by jxom** [comment]
Does deployless multicall work when calls fail, and `allowFailure` is enabled? Seems like the tests around those in `multicall.test.ts` seem to fail hard. If it doesn't support, maybe we should not allow `allowFailure` to be enabled.

**Comment by 0xRapid** [comment]
Would love to see that merged soon! @not-reed friendly bump

**Comment by not-reed** [comment]
@0xRapid  agreed :raised_hands: 

@jxom You are correct, the original deployless multicall3 contract seems to not handle this properly and I think behaves more like multicall2 in this capacity. not allowing `allowFailure` seems reasonable. I was hoping I could get some time to fix the contract bytecode itself, but haven't been able to get there yet.

**Comment by jxom** [comment]
Closing for now as this version of deployless multicall doesn't work well with reverts. Happy to accept a PR with an updated version that supports it!

**Pull request closed without merging**